File: GGAndroid_2.0/GGCore/GGVMCore/src/main/java/io/ganguo/vmodel/ViewModelDialog.java
package io.ganguo.vmodel;

import android.content.Context;
import android.databinding.ViewDataBinding;

import io.ganguo.library.ui.dialog.BaseDialog;
import io.ganguo.utils.util.log.Logger;
import io.ganguo.vmodel.view.ViewModelInterface;

/**
 * Created by Roger on 7/13/16.
 */
public abstract class ViewModelDialog<T extends ViewDataBinding, B extends BaseViewModel> extends BaseDialog
        implements BaseViewModel.OnViewAttachListener<B>, ViewModelInterface<B> {

    private B mViewModel;

    @Override
    public boolean checkViewModel() {
        boolean isNotNull = getViewModel() != null;
        if (!isNotNull) {
            Logger.e("viewModel is null");
        }
        return isNotNull;
    }

    public ViewModelDialog(Context context) {
        super(context);
    }

    public ViewModelDialog(Context context, int theme) {
        super(context, theme);
    }

    @Override
    public void beforeInitView() {
        ViewModelHelper.bind(this, getViewModel());
        setContentView(getViewModel().getRootView());
        getViewModel().setOnViewAttachListener(this);
    }

    @Override
    public void initView() {
    }

    @Override
    public void initListener() {
    }

    @Override
    public void initData() {

    }

    @SuppressWarnings("unchecked raw type")
    public T getBinding() {
        return (T) getViewModel().getView().getBinding();
    }

    @Override
    public B getViewModel() {
        if (mViewModel == null) {
            mViewModel = createViewModel();
        }
        return mViewModel;
    }

    /**
     * Bind ViewModel Lifecycle
     */
    @Override
    protected void onStop() {
        super.onStop();
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onStop();
        }
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (checkViewModel()) {
            getViewModel().getLifecycleHelper().onDestroy();
        }
    }

    /**
     * Create ViewModel here
     * if you want to getViewModel instance use {@link #getViewModel()} instead this
     */
    @Override
    public abstract B createViewModel();

    @Override
    protected boolean isFullScreen() {
        return false;
    }
}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/core/view/ViewModelDialog.kt
package io.ganguo.viewmodel.core.view

import android.content.Context
import androidx.databinding.ViewDataBinding
import io.ganguo.core.R
import io.ganguo.core.ui.dialog.BaseDialog
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.BaseViewModel.OnViewAttachListener
import io.ganguo.viewmodel.core.ViewModelHelper
import io.ganguo.viewmodel.core.creator.ViewModelCreator

/**
 * <pre>
 *     @author : leo
 *     time   : 2020/05/29
 *     desc   : Base ViewModelDialog
 * </pre>
 * @param T : ViewDataBinding
 * @param B : BaseViewModel<*>
 * @property viewModel B
 * @constructor
 */
abstract class ViewModelDialog<T : ViewDataBinding, B : BaseViewModel<*>>(context: Context, theme: Int=R.style.Dialog_Immersion)
    : BaseDialog(context, theme), OnViewAttachListener<B>, ViewModelCreator<B> {
    override val viewModel: B by lazy { createViewModel() }

    override fun beforeInitView() {
        ViewModelHelper.bind(this, viewModel)
        setContentView(viewModel.rootView)
    }

    override fun initView() {}
    override fun initListener() {
        viewModel.onViewAttachListener = this as OnViewAttachListener<BaseViewModel<*>>
    }

    override fun initData() {}


    /**
     * Bind ViewModel Lifecycle
     */
    override fun onStop() {
        super.onStop()
        viewModel.lifecycleHelper.onLifecycleStop()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        viewModel.lifecycleHelper.onLifecycleDestroy()
    }

    /**
     * Create ViewModel here
     * if you want to getViewModel instance use [.getViewModel] instead this
     */
    abstract override fun createViewModel(): B
}